## Some ideas for your final project


- *Surrogate variable analysis*
    - Leek, J. T. & Storey, J. D. Capturing Heterogeneity in Gene Expression Studies by Surrogate Variable Analysis. PLoS Genet 3, e161 (2007). **Group: Anna Engstrom, Xieting Zhang and Yu Yin**

- *RNA-seq*
    - Anders, S., Reyes, A. & Huber, W. Detecting differential usage of exons from RNA-seq data. Genome Res. 22, 2008-2017 (2012). **Group: Anna Plantinga, Katie Wilson, Shih-Yu (Shirley) Chang**
    - Frazee, A. C., Sabunciyan, S., Hansen, K. D., Irizarry, R. A. & Leek, J. T. Differential expression analysis of RNA-seq data at single-base resolution. Biostatistics kxt053 (2014). doi:10.1093/biostatistics/kxt053. 
    - Risso, D., Schwartz, K., Sherlock, G. & Dudoit, S. GC-content normalization for RNA-Seq data. BMC Bioinformatics 12, 480 (2011). **Group: David Whitney, Jun Wung and Caitlin McHugh**
    - Katz, Y., Wang, E. T., Airoldi, E. M. & Burge, C. B. Analysis and design of RNA sequencing experiments for identifying isoform regulation. Nat. Methods 7, 1009-1015 (2010). **Group: Elisabeth Rosenthal and Xieting Zhang and Yu Yin**
    - Hansen, K. D., Irizarry, R. A. & Wu, Z. Removing technical variability in RNA-seq data using conditional quantile normalization. Biostatistics 13, 204-216 (2012).
    - Trapnell, C. et al. Differential gene and transcript expression analysis of RNA-seq experiments with TopHat and Cufflinks. Nat Protoc 7, 562-578 (2012). **Group: Ravi Sood, David Scoville and Yanfei Li**
    
- *eQTL/sQTL*
    - Zhao, K., Lu, Z.-X., Park, J. W., Zhou, Q. & Xing, Y. GLiMMPS: robust statistical model for regulatory variation of alternative splicing using RNA-seq data. Genome Biol. 14, R74 (2013).
    
- *Single-cell gene-expression*
    - McDavid, A. et al. Data exploration, quality control and testing in single-cell qPCR-based gene expression experiments. Bioinformatics 29, 461-467 (2013).
    - Shalek, A. K. et al. Single-cell transcriptomics reveals bimodality in expression and splicing in immune cells. Nature 498, 236-240 (2013). 

- *Gene expression deconvolution*
    - Shen-Orr, S. S. et al. Cell type-specific gene expression differences in complex tissues. Nat. Methods 7, 287-289 (2010). **Group: Jia Jin Kee and Ju Young Park**
    
- *ChIP-seq*
    - Zhang, Y. et al. Model-based analysis of ChIP-Seq (MACS). Genome Biol. 9, R137 (2008).

- *Methylation*
    - Hansen, K. D., Langmead, B. & Irizarry, R. A. BSmooth: from whole genome bisulfite sequencing reads to differentially methylated regions. Genome Biol. (2012).
    - Saito, Y., Tsuji, J. & Mituyama, T. Bisulfighter: accurate detection of methylated cytosines and differentially methylated regions. Nucleic Acids Res. gkt1373 (2014). doi:10.1093/nar/gkt1373. **Group: YuChi Chang and Travis, Kenneth Chen**


## What you will have to do
1. Read and understand the paper
2. Reproduce some of the analysis
3. Summarize the paper with the analysis in reproducible presentation (a .Rpres file)
4. Present your work to the rest of the class
